woocommerce/woocommerce-android
# [REST API] Tracking

Closes: #8190


### Description
This PR adds tracking to the site credentials login and application passwords usage according to this plan pe5sF9-13h-p2 

For some tracking events, I needed two preparatory changes, and this what made the PR bigger:
1. [57af766](https://github.com/woocommerce/woocommerce-android/pull/8220/commits/57af7660e2a9791f66c8afb264b116d0f441af16): to expose the error when fetching user role fails.
2. [e9ebed7](https://github.com/woocommerce/woocommerce-android/pull/8220/commits/e9ebed7b610df2154e2dfae7e0da86833cd76a27): to expose events about application password generation failures.

For the rest, since some events were added in previous PRs, I prepared this table to help you follow where each event was added:

| Events | Commit |
| --- | --- |
| login_site_address_site_info_requested <br /> login_site_address_site_info_succeeded | 812badeec85fbb0f24a7ed86661b3dfca92be859 |
| unified_login_step (login_store_creds) <br/> unified_login_interaction (submit)  | 67885531fd878f7b7cc3ade537640858c2925b15 (previous PR) |
| unified_login_failure <br /> login_site_credentials_login_failed | 306b8d76f935ed028ba5f46a49db0c965d6eeb64 |
| unified_login_step (not_woo_store) |  b57fbaef54c012c333e64813f1a6b04f5f11bddd (previous PR) |
| login_insufficient_role | f6547ff107cdd12bb1f2994c274554ef6ebf839c |
| rest_api_login_experiment | 87475d00cfc2daac54766cffabe74b7bd88b2564 (previous PR) |
| signed_in |  ac3c712517c69ee8eec7a391572dc5e2469bbe5f |
| application_passwords_new_password_created <br/> application_passwords_generation_failed | b984f38094f813cdbd584f0d325cd05162468671 |

### Testing instructions
I tried to have detailed test scenarios below because they helped me validate the events, I don't think it's expected to test all those scenarios in a code review, just test what you can, then review the code and check the tests added in [88a2d3a](https://github.com/woocommerce/woocommerce-android/pull/8220/commits/88a2d3a1656f50ab27413c878a0ac1a6bb3f1ca2)

To test error scenarios below, it's better to use a proxy app (like Charles), and you might find a website that simulates different status codes like https://httpstat.us/ useful:

1. Make sure you are assigned the Treatment variant of the REST API experiment (either by [hardcoding](https://github.com/woocommerce/woocommerce-android/blob/9841083ed877c1846ca66d43a44841ecfb320837/WooCommerce/src/main/kotlin/com/woocommerce/android/experiment/RESTAPILoginExperiment.kt#L43) it or by setting a test device)
2. Open the app, then click on "Enter site address"
3. Enter the site address.
4. In the next screen, enter wrong site credentials.
5. Confirm the following events are sent:
```
unified_login_interaction, Properties: {"flow":"login_store_creds","step":"username_password", "click":"submit"...}
unified_login_failure, Properties: {"flow":"login_store_creds","step":"username_password" ...}
login_site_credentials_login_failed, Properties: { "step":"authentication" ... }
```
6. Simulate a site that has Application Passwords disabled (`wp/v2/users/me/application-passwords` should return an error that contains the returns the body: `{ "code" : "application_passwords_disabled" }`), or alternatively install a plugin that disable them.
7. Attempt login again but using correct credentials.
8. Confirm the following event is sent: `application_passwords_generation_failed, Properties: {"cause":"feature_disabled","scenario":"generation",...}
9. Re-enable application passwords.
10. Simulate a generic error for `wp/v2/users/me/application-passwords`
11. Attempt login again.
12. Confirm the following events are sent:
```
application_passwords_generation_failed, Properties: {"cause":"other","scenario":"generation"...}
login_site_credentials_login_failed, Properties: {"step":"application_password_generation"...}
```
13. Remove the error of `wp/v2/users/me/application-passwords`, then simulate a generic error for `/wc/v3/settings`
14. Attempt login again.
15. Confirm the following event are sent: 
```
application_passwords_new_password_created: Properties: { "scenario":"generation"...}
login_site_credentials_login_failed, Properties: {"step":"woo_status"...}
```
16. Remove the error of `/wc/v3/settings`, then simulate a generic error for `/wp/v2/users/me`
17. Confirm the following event is sent: `login_site_credentials_login_failed, Properties: {"step":"user_role"...}`
18. Remove the error simulation.
19. Attempt login.
20. Confirm the following event is sent: `signed_in`
21. Go to your site, then change your username's password and revoke the application's password.
22. Trigger any API requests in the app.
23. Confirm the following event is sent: `application_passwords_generation_failed, Properties: {"cause":"authorization_failed","scenario":"regeneration"...}`
